src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_aggregate.rs

```diff
@@ -0,0 +1,156 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::optimizer::rule::Rule;
+use crate::optimizer::rule::TransformResult;
+use crate::optimizer::RuleID;
+use crate::optimizer::SExpr;
+use crate::plans::Aggregate;
+use crate::plans::AggregateMode;
+use crate::plans::Filter;
+use crate::plans::PatternPlan;
+use crate::plans::RelOp;
+use crate::plans::RelOp::Aggregate as OpAggregate;
+use crate::plans::RelOp::Pattern;
+use crate::plans::RelOperator;
+
+/// Heuristic optimizer runs in a bottom-up recursion fashion. If we match a plan like
+/// Filter-Aggregate-* and push down filter to Filter(Optional)-Aggregate-Filter-*, this will not
+/// work. RuleSplitAggregate will be applied first, since it's bottom up, then this rule, which
+/// cause the plan be like Filter(Optional)-Aggregate-Filter-Aggregate-*, which makes no sense.
+/// Hence we match 2 bundled Aggregate Ops:
+///
+/// Input:  Filter
+///           \
+///          Aggregate(Final)
+///            \
+///          Aggregate(Partial)
+///             \
+///              *
+///
+/// Output: Filter(Optional)
+///           \
+///          Aggregate(Final)
+///             \
+///            Aggregate(Partial)
+///               \
+///              Filter
+///                \
+///                 *
+pub struct RulePushDownFilterAggregate {
+    id: RuleID,
+    pattern: SExpr,
+}
+
+impl RulePushDownFilterAggregate {
+    pub fn new() -> Self {
+        Self {
+            id: RuleID::PushDownFilterAggregate,
+            pattern: SExpr::create_unary(
+                PatternPlan {
+                    plan_type: RelOp::Filter,
+                }
+                .into(),
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: OpAggregate,
+                    }
+                    .into(),
+                    SExpr::create_unary(
+                        PatternPlan {
+                            plan_type: OpAggregate,
+                        }
+                        .into(),
+                        SExpr::create_leaf(PatternPlan { plan_type: Pattern }.into()),
+                    ),
+                ),
+            ),
+        }
+    }
+}
+
+impl Rule for RulePushDownFilterAggregate {
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(&self, s_expr: &SExpr, state: &mut TransformResult) -> common_exception::Result<()> {
+        let filter: Filter = s_expr.plan().clone().try_into()?;
+        if filter.is_having {
+            let agg1 = s_expr.child(0)?;
```

Comment:
It would be better to change the names: `agg1`|`agg2` to a meaningful name :)

Reply:
ack

Reply:
done!